glusterd2/store/store.go

```diff
@@ -111,9 +113,65 @@ func New(conf *Config) (*GDStore, error) {
 	if err = store.publishLiveness(); err != nil {
 		return nil, err
 	}
+
+	store.stop = make(chan struct{})
+
+	go store.keepSessionLeaseAlive()
 	return store, nil
 }
 
+// keepSessionLeaseAlive keeps session lease alive forever. If etcd goes down
+// and comes up then existing session lease will not be refreshed and expires after
+// lease ttl duration. keepSessionLeaseAlive checks session lease information on a
+// regular interval. It will start refreshing session lease if it is not getting
+// refreshed.
+func (s *GDStore) keepSessionLeaseAlive() {
+	var (
+		ticker         = time.NewTicker(time.Second * 5)
+		printedFailure bool
+	)
+
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-s.stop:
+			return
+		case <-ticker.C:
+			ctx, cancel := context.WithTimeout(context.Background(), getTimeout*time.Second)
+			defer cancel()
```

Comment:
are you sure will this defer gets called inside infinite for loop?

Reply:
@Madhu-1 defer will get called after control came out of the loop, i.e when GDstore will get closed,
but since the context object (ctx) get used at only one place so I have removed the defer.